sound/soc/codecs/rt715-sdca.c

```diff
@@ -0,0 +1,935 @@
+	unsigned int tmp;
+	int ret;
+
+	ret = rt715_index_read(rt715, nid, reg, &tmp);
+	if (ret < 0)
+		return ret;
+
+	set_mask_bits(&tmp, mask, val);
+
+	return rt715_index_write(rt715, nid, reg, tmp);
+}
+
+/* SDCA Volume/Boost control */
+static int rt715_set_amp_gain_put_sdca(struct snd_kcontrol *kcontrol,
+		struct snd_ctl_elem_value *ucontrol)
+{
+	struct snd_soc_component *component = snd_kcontrol_chip(kcontrol);
+	struct soc_mixer_control *mc =
+		(struct soc_mixer_control *)kcontrol->private_value;
+	struct rt715_sdca_priv *rt715 = snd_soc_component_get_drvdata(component);
+	unsigned int val_l, val_r, gain_l_val, gain_r_val;
+	unsigned int i;
+
+	/* control value to 2s complement */
+	/* L channel */
+	gain_l_val = ucontrol->value.integer.value[0];
+	if (gain_l_val > mc->max)
+		gain_l_val = mc->max;
+	val_l = gain_l_val;
+
+	if (mc->shift == 8)
+		gain_l_val = (gain_l_val * 10) << mc->shift;
+	else {
+		gain_l_val = abs(gain_l_val - mc->shift) * 375;
+		gain_l_val <<= 8;
+		gain_l_val /= 1000;
+		if (val_l <= mc->shift) {
+			gain_l_val = ~gain_l_val;
+			gain_l_val += 1;
+		}
+		gain_l_val &= 0xffff;
+	}
+
+	/* R channel */
+	gain_r_val = ucontrol->value.integer.value[1];
+	if (gain_r_val > mc->max)
+		gain_r_val = mc->max;
+	val_r = gain_r_val;
+
+	if (mc->shift == 8)
+		gain_r_val = (gain_r_val*10) << mc->shift;
+	else {
+		gain_r_val = abs(gain_r_val - mc->shift) * 375;
+		gain_r_val <<= 8;
+		gain_r_val /= 1000;
+		if (val_r <= mc->shift) {
+			gain_r_val = ~gain_r_val;
+			gain_r_val += 1;
+		}
+		gain_r_val &= 0xffff;
+	}
+
+	for (i = 0; i < 3; i++) { /* retry 3 times at most */
+		if (gain_l_val == gain_r_val) {
+			/* Set both L/R channels at the same time */
+			regmap_write(rt715->regmap,
+				mc->reg, gain_l_val);
+			regmap_write(rt715->regmap,
+				mc->rreg, gain_l_val);
+		} else {
+			/* Lch*/
+			regmap_write(rt715->regmap,
+				mc->reg, gain_l_val);
+			/* Rch */
+			regmap_write(rt715->regmap,
+				mc->rreg, gain_r_val);
+		}
+
+		regmap_read(rt715->regmap, mc->reg, &val_l);
+		regmap_read(rt715->regmap, mc->rreg, &val_r);
+		if (val_r == gain_r_val && val_l == gain_l_val)
+			break;
+	}
+
+	return 0;
+}
+
+static int rt715_set_amp_gain_get_sdca(struct snd_kcontrol *kcontrol,
+		struct snd_ctl_elem_value *ucontrol)
+{
+	struct snd_soc_component *component = snd_kcontrol_chip(kcontrol);
+	struct soc_mixer_control *mc =
+		(struct soc_mixer_control *)kcontrol->private_value;
+	struct rt715_sdca_priv *rt715 = snd_soc_component_get_drvdata(component);
+	unsigned int val_l, val_r, ctl_l, ctl_r, neg_flag = 0;
+
+	regmap_read(rt715->regmap, mc->reg, &val_l);
+	regmap_read(rt715->regmap, mc->rreg, &val_r);
+
+	/* L channel */
+	if (mc->shift == 8) {
+		ctl_l = (val_l >> mc->shift) / 10;
+	} else {
+		ctl_l = val_l;
+		if (ctl_l & BIT(15)) {
+			ctl_l = ~(val_l - 1) & 0xffff;
+			neg_flag = 1;
+		}
+		ctl_l *= 1000;
+		ctl_l >>= 8;
+		if (neg_flag)
+			ctl_l = mc->shift - ctl_l / 375;
+		else
+			ctl_l = mc->shift + ctl_l / 375;
```

Comment:
375 is a magic number ?  I find many of it. Can we define a macro to make it more instructive ?